### PR TITLE
Controller ready

### LINKED ISSUE #71: Creating a new "model_ready" event in widget.js
I encountered the following problem with widgets that create sub-widgets from the JavaScript side in their JavaScript model constructor: 

For example the gamepad widget creates subwidgets for the buttons and axes in its constructor. Once the `connected` trait attribute is set to True, the user may wire things to the gamepad buttons...

However, in the case of a page reload, I would like these widgets **not** to be replaced by new ones, when the constructor of the JavaScript side is executed. The part of the constructor containing the creation of the children should only occur after the state request resolves.

One way to solve this issue would be to create a `model-ready` event for that, of have a certain promise that gets resolved right away in the case of a new model, and after the state request resolves in case of page reload.

Suggestions welcome.

@jdfreder @jasongrout

### PR DESCRIPTION
**Pull request by SylvainCorlay**
Supersedes #183 but rebased on master. -- closes #71.
Comes on the top of #190.

### CODE REVIEW

**Comment by jdfreder** [comment]
closes https://github.com/ipython/ipywidgets/issues/71

File: ipywidgets/static/widgets/js/widget_controller.js [hunk 1]
```diff
@@ -144,12 +155,13 @@ define([
         },
 
         update_loop: function() {
-            /* Populates axes and button values, until the gamepad is disconnected.
+            /* Upadte axes and buttons values, until the gamepad is disconnected.
```
Comment (jdfreder) [inline ipywidgets/static/widgets/js/widget_controller.js]: typo `update`

**Comment by jdfreder** [comment]
One small comment typo, otherwise this is ready to go!

**Comment by SylvainCorlay** [comment]
Except that it does not work. Working on it.

**Comment by jdfreder** [comment]
Oh, missed the "WIP" note in the title

**Comment by SylvainCorlay** [comment]
Lol. I think I found a problem in ipykernel. Opened a related PR in ipykernel.

**Comment by jdfreder** [comment]
Oh okay, good.  Sorry you had to dig that deep, sounds like a pain.

**Comment by jdfreder** [comment]
In the current state, does this PR break anything in master more than it's already broken, without https://github.com/ipython/ipykernel/pull/67 merged?  I'm wondering if we can merge this for 4.1 or if we should push on without it.

**Comment by SylvainCorlay** [comment]
I don't think that it breaks anything but persistence is not quite there yet for the controller. I think that there is a problem with the beastmode code.

**Comment by SylvainCorlay** [comment]
I don't think it breaks anything but persistence is not there yet for the controller widget.
There is a problem with the beastmode code. Now looking into it.

**Comment by jdfreder** [comment]
@SylvainCorlay does your latest commit (ee6d590) fix the problem?

**Comment by SylvainCorlay** [comment]
This seems to be ok now.

**Comment by SylvainCorlay** [comment]
Now it seems to be working fine.

**Comment by jdfreder** [comment]
Woohoo!

**Pull request merged**